File: 53. Maximum Subarray/53.cpp
/*
dynamic programming: dp[i] maximum subarray sum ending at index i
time: O(N) 
space: O(N)
*/
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        vector<int> dp(nums.size());
        dp[0] = nums[0];
        int maxsum = dp[0];

        for(int i =1; i< nums.size(); i++){
            dp[i] = max(nums[i], nums[i] + dp[i-1]);
            maxsum = max(maxsum, dp[i]);
        }

        return maxsum;

    }
};

File: 53. Maximum Subarray/53.py
"""
time: O(n)
space: O(1)
"""
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        bestSum = nums[0]
        currentSum = nums[0]

        for i in range(1, len(nums)):
            currentSum = max(nums[i], currentSum + nums[i])
            bestSum = max(currentSum, bestSum)

        return bestSum
"""
sliding window? -> no because need a specific criteria to make subarray valid -> too many possibilities
maximum subarray problem -> kadane's algorithm dp
"""
